build.rb

```diff
@@ -10,21 +10,44 @@
   'lang:fr_CA' => 'fr',
 }.freeze
 
-query = <<~SPARQL
+legislative_dir = Pathname.new(__FILE__).dirname.join('legislative')
+index_file = legislative_dir.join('index.json')
+
+def date_condition(start_date, end_date)
+  return '' unless start_date && end_date
+  <<~DATE_CONDITION
+    BIND(COALESCE(?start, "1000-01-01T00:00:00Z"^^xsd:dateTime) AS ?start_or_sentinel)
+    BIND(COALESCE(?end, "9999-12-31T00:00:00Z"^^xsd:dateTime) AS ?end_or_sentinel)
+    FILTER (?end_or_sentinel >= "#{start_date}"^^xsd:dateTime)
+    FILTER (?start_or_sentinel <= "#{end_date}"^^xsd:dateTime)
+DATE_CONDITION
+end
+
+def term_condition(term_item_id)
+  return '' unless term_item_id
+  "?statement pq:P2937 wd:#{term_item_id} ."
+end
+
+def query(position_item_id:, term_item_id: nil, start_date: nil, end_date: nil, **_)
+  unless !!term_item_id ^ !!(start_date and end_date)
```

Comment:
Are the double bangs necessary here? I've had to peer extra carefully at this for quite a while to try to work out if there's something subtle going on.

Reply:
I wanted xor, but `^` only seems to work with integers or booleans; it doesn't do an implicit conversion of anything truthy or falsy to boolean:

```
irb(main):001:0> true ^ true
=> false
irb(main):002:0> 1 ^ 1
=> 0
irb(main):003:0> 'foo' ^ 'bar'
NoMethodError: undefined method `^' for "foo":String
	from (irb):3
	from /home/mark/.rbenv/versions/2.4.2/bin/irb:11:in `<main>'
```

... so the double-bangs are effectively casting to boolean.